tests/generators/generators.js

```diff
@@ -0,0 +1,107 @@
+function *stmt_yield(): Generator<number, void, void> {
+  yield 0; // ok
+  yield ""; // error: string ~> number
+}
+
+function *stmt_next(): Generator<void, void, number> {
+  (yield undefined : number); // ok
+  (yield undefined : string); // error: number ~> string
+}
+
+function *stmt_return_ok(): Generator<void, number, void> {
+  return 0; // ok
+}
+
+function *stmt_return_err(): Generator<void, number, void> {
+  return ""; // error: string ~> number
+}
+
+function *infer_stmt() {
+  var x: boolean = yield 0;
+  return "";
+}
+for (var x of infer_stmt()) { (x : string) } // error: number ~> string
+var infer_stmt_next = infer_stmt().next(0).value; // error: number ~> boolean
+if (typeof infer_stmt_next === "undefined") {
+} else if (typeof infer_stmt_next === "number") {
+} else {
+  (infer_stmt_next : boolean) // error: string ~> boolean
+}
+
+declare function *decl(): Generator<number, string, boolean>;
+for (var x of decl()) { (x : string) } // error: number ~> string
+var decl_next = decl().next(0).value; // error: number ~> boolean
+if (typeof decl_next === "undefined") {
+} else if (typeof decl_next === "number") {
+} else {
+  (decl_next : boolean) // error: string ~> boolean
+}
+
+function *widen_next() {
+  var x = yield 0;
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+widen_next().next(0)
+widen_next().next("")
+widen_next().next(true)
+
+function *widen_yield() {
+  yield 0;
+  yield "";
+  yield true;
+}
+for (var x of widen_yield()) {
+  if (typeof x === "number") {
+  } else if (typeof x === "boolean") {
+  } else {
+    (x : string) // ok, sherlock
+  }
+}
+
+function *delegate_next_generator() {
+  function *inner() {
+    var x: number = yield undefined; // error: string ~> number
+  }
+  yield *inner();
+}
+delegate_next_generator().next("");
+
+function *delegate_yield_generator() {
+  function *inner() {
+    yield "";
+  }
+
+  yield *inner();
+}
+for (var x of delegate_yield_generator()) {
+  (x : number) // error: string ~> number
+}
+
+function *delegate_return_generator() {
+  function *inner() {
+    return "";
+  }
+
+  var x: number = yield *inner() // TODO: error: string ~> number
+}
+
+// only generators can make use of a value passed to next
+function *delegate_next_iterable(xs: Array<number>) {
+  yield *xs;
+}
+delegate_next_iterable([]).next(""); // TODO: error: string ~> void
```

Comment:
Here flow gives an unexpected error, instead of the expected error:

```
generators.js:94:3,11: yield* delegate
Error:
[LIB] core.js:306:5,40: property next
Property not found in
generators.js:94:10,11: array type
```

It seems like the [first case of the `$yieldDelegate` intersection](https://github.com/facebook/flow/pull/534/files#diff-f265d6a6b6b64d4d096fddcf7d620a76R310) is failing instead of failing over to the second case, which should pass. It's also very likely I don't understand the speculative match failure stuff.